addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/handler/StatefulHandlerCallback.java

```diff
@@ -0,0 +1,159 @@
+import org.eclipse.smarthome.core.types.State;
+
+/**
+ * Defines an implementation of {@link NeeoHandlerCallback} that will remember the last state
+ * for an channelId and suppress the callback if the state hasn't changed.
+ *
+ * @author Tim Roberts - initial contribution
+ */
+class StatefulHandlerCallback implements NeeoHandlerCallback {
+
+    /** The wrapped callback. */
+    private final NeeoHandlerCallback wrappedCallback;
+
+    /** The state by channel id. */
+    private final Map<String, State> state = new ConcurrentHashMap<String, State>();
+
+    /** The status lock. */
+    private final Lock statusLock = new ReentrantLock();
+
+    /** The last thing status. */
+    private ThingStatus lastThingStatus = null;
+
+    /** The last thing status detail. */
+    private ThingStatusDetail lastThingStatusDetail = null;
+
+    /**
+     * Create the callback from the other {@link NeeoHandlerCallback}.
+     *
+     * @param wrappedCallback a non-null {@link NeeoHandlerCallback}
+     * @throws IllegalArgumentException if wrappedCallback is null
+     */
+    public StatefulHandlerCallback(NeeoHandlerCallback wrappedCallback) {
+        Objects.requireNonNull(wrappedCallback, "wrappedCallback cannot be null");
+
+        this.wrappedCallback = wrappedCallback;
+    }
+
+    /**
+     * Overrides the status changed to simply call the {@link #wrappedCallback}.
+     *
+     * @param status the new status
+     * @param detail the new detail
+     * @param msg the new message
+     */
+    @Override
+    public void statusChanged(ThingStatus status, ThingStatusDetail detail, String msg) {
+        statusLock.lock();
+        try {
+            // Simply return we match the last status change (prevents loops if changing to the same status)
+            if (status == lastThingStatus && detail == lastThingStatusDetail) {
+                return;
+            }
+
+            lastThingStatus = status;
+            lastThingStatusDetail = detail;
+        } finally {
+            statusLock.unlock();
+        }
+
+        // If we got this far - call the underlying one
+        wrappedCallback.statusChanged(status, detail, msg);
+    }
+
+    /**
+     * Overrides the state changed to determine if the state is new or changed and then
+     * to call the {@link #wrappedCallback} if it has.
+     *
+     * @param channelId the channel id that changed
+     * @param newState the new state
+     */
+    @Override
+    public void stateChanged(String channelId, State newState) {
+        if (StringUtils.isEmpty(channelId)) {
+            return;
+        }
+
+        final State oldState = state.get(channelId);
+
+        if (oldState == null && newState == null) {
+            return;
+        }
+
+        // If they are equal - nothing changed
+        if (oldState != null && oldState.equals(newState)) {
+            return;
+        }
+
+        if (newState != null && newState.equals(oldState)) {
```

Comment:
The contract of equals states it should work both ways. Thus reversing the variables should not give a different result. Meaning this check is dead code, as one of the above will already have returned or if `newState !=null`, but `oldstate == null` they are not equal and thus won't return here.